### PR TITLE
fix some races

### PR DESCRIPTION
**Pull request by roblaszczak**

### CODE REVIEW

File: internal/channel.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package internal
+
+// IsChannelClosed returns true if provided `chan struct{}` is closed.
+// IsChannelClosed panics if message is sent to this channel.
+func IsChannelClosed(channel chan struct{}) bool {
+	select {
+	case _, ok := <-channel:
+		if ok {
+			panic("received unexpected message")
```
[the remark this answers is not in the document]
Comment (roblaszczak) [reply internal/channel.go]: it wasn't panicking at the beginning but after thinking a bit I assumed that you will not anyway send any `struct{}` message to this channel - it should be used only for closing, and if you will something went really wrong

File: message/infrastructure/nats/subscriber.go [hunk 2]
```diff
@@ -391,3 +394,10 @@ func (s *StreamingSubscriber) Close() error {
 
 	return result
 }
+
+func (s *StreamingSubscriber) isClosed() bool {
+	s.subsLock.RLock()
+	defer s.subsLock.RUnlock()
+
+	return s.closed
```
Comment (bkielbasa) [inline message/infrastructure/nats/subscriber.go]: can you explain why you added the rmutex here?

Comment (roblaszczak) [reply message/infrastructure/nats/subscriber.go]: `s.closed` can be concurrently read in `processMessage` and written in `Close()` during this, and we have a race condition

File: message/infrastructure/googlecloud/publisher.go [hunk 3]
```diff
@@ -87,7 +95,13 @@ func (p *Publisher) Publish(topic string, messages ...*message.Message) error {
 		return err
 	}
 
+	logFields := make(watermill.LogFields, 2)
+	logFields["topic"] = topic
+
 	for _, msg := range messages {
+		logFields["message_uuid"] = msg.UUID
+		p.config.Logger.Trace("Sending message to Google Pub/Sub", logFields)
```
[the remark this answers is not in the document]
Comment (roblaszczak) [reply message/infrastructure/googlecloud/publisher.go]: changed

**Pull request merged**